akka-persistence/src/main/scala/akka/persistence/Persistent.scala

```diff
@@ -38,6 +38,15 @@ object AtomicWrite {
 }
 
 final case class AtomicWrite(payload: immutable.Seq[PersistentRepr]) extends PersistentEnvelope with Message {
+
+  // TODO: quite costy... do we need to do this? We only internally create AWs hm...
+  require(!payload.exists(_.persistenceId != payload.head.persistenceId),
+    s"AtomicWrite must contain messages for the same persistenceId, yet different persistenceIds found: ${payload.map(_.persistenceId).toSet}")
+
+  def persistenceId = payload.head.persistenceId
+  def lowestSequenceNr = payload.map(_.sequenceNr).min // this assumes they're gapless; they should be
```

Comment:
in theory should also guard against empty seqs... it's a question of how much we trust only us creating these I guess.